master fails to build due to '...is ambiguuous' error

Getting an error trying to  build 3084a57 from source; `master` built fine as of a few days ago in my environment.

Environment info:
```
malmaud@devon:~/pytorch$ gcc --version
gcc (Ubuntu 5.4.0-6ubuntu1~16.04.6) 5.4.0 20160609

malmaud@devon:~/pytorch$ lsb_release -a
No LSB modules are available.
Distributor ID: Ubuntu
Description:    Ubuntu 16.04.2 LTS
Release:        16.04
Codename:       xenial
```

Error when running `python setup.py install`:

```
[1/3] Building CXX object CMakeFiles/THD.dir/base/data_channels/DataChannelMPI.cpp.o
FAILED: CMakeFiles/THD.dir/base/data_channels/DataChannelMPI.cpp.o
/usr/bin/c++  -DWITH_CUDA=1 -DWITH_GLOO=1 -DWITH_MPI=1 -D_THD_CORE=1 -I/storage/malmaud/pytorch/torch/lib/tmp_install/include -I/storage/malmaud/cuda-9.1/include -I/storage/malmaud/pytorch/torch/lib/THD/CUDA_SDK_ROOT_DIR-NOTFOUND/common/inc -I/storage/malmaud/pytorch/torch/lib/THD -I/usr/lib/openmpi/include/openmpi/opal/mca/event/libevent2021/libevent -I/usr/lib/openmpi/include/openmpi/opal/mca/event/libevent2021/libevent/include -I/usr/lib/openmpi/include -I/usr/lib/openmpi/include/openmpi -DTH_INDEX_BASE=0 -I"/storage/malmaud/pytorch/torch/lib/tmp_install/include"   -I"/storage/malmaud/pytorch/torch/lib/tmp_install/include/TH" -I"/storage/malmaud/pytorch/torch/lib/tmp_install/include/THC"   -I"/storage/malmaud/pytorch/torch/lib/tmp_install/include/THS" -I"/storage/malmaud/pytorch/torch/lib/tmp_install/include/THCS"   -I"/storage/malmaud/pytorch/torch/lib/tmp_install/include/THNN" -I"/storage/malmaud/pytorch/torch/lib/tmp_install/include/THCUNN" -DOMPI_SKIP_MPICXX=1 -fexceptions  -std=c++11 -std=c++11 -O3 -DNDEBUG -fPIC -MD -MT CMakeFiles/THD.dir/base/data_channels/DataChannelMPI.cpp.o -MF CMakeFiles/THD.dir/base/data_channels/DataChannelMPI.cpp.o.d -o CMakeFiles/THD.dir/base/data_channels/DataChannelMPI.cpp.o -c /storage/malmaud/pytorch/torch/lib/THD/base/data_channels/DataChannelMPI.cpp
/storage/malmaud/pytorch/torch/lib/THD/base/data_channels/DataChannelMPI.cpp: In member function ‘virtual void thd::DataChannelMPI::allGather(std::vector<at::Tensor>&, at::Tensor&, THDGroup)’:
/storage/malmaud/pytorch/torch/lib/THD/base/data_channels/DataChannelMPI.cpp:194:34: error: conversion from ‘std::size_t {aka long unsigned int}’ to ‘at::Scalar’ is ambiguous
     output[i].copy_(recv_buffer[i]);
                                  ^
In file included from /storage/malmaud/pytorch/torch/lib/tmp_install/include/ATen/ATen.h:5:0,
                 from /storage/malmaud/pytorch/torch/lib/THD/base/data_channels/../ChannelUtils.hpp:3,
                 from /storage/malmaud/pytorch/torch/lib/THD/base/data_channels/../DataChannel.hpp:4,
                 from /storage/malmaud/pytorch/torch/lib/THD/base/data_channels/DataChannelMPI.hpp:3,
                 from /storage/malmaud/pytorch/torch/lib/THD/base/data_channels/DataChannelMPI.cpp:1:
```

+1

```
pytorch/torch/lib/THD/base/data_channels/DataChannelMPI.cpp: In member function 'virtual void thd::DataChannelMPI::allGather(std::vector<at::Tensor>&, at::Tensor&, THDGroup)':
/data/sls/u/sameerk/repos/pyro-forked/pyro/examples/dmm-speech/pytorch/torch/lib/THD/base/data_channels/DataChannelMPI.cpp:194:34: error: conversion from 'std::size_t {aka long unsigned int}' to 'at::Scalar' is ambiguous
     output[i].copy_(recv_buffer[i]);
```

+1 output:
```
/home/niutr/pytorch/torch/lib/THD/base/data_channels/DataChannelMPI.cpp: In member function ‘virtual void thd::DataChannelMPI::allGather(std::vector<at::Tensor>&, at::Tensor&, THDGroup)’:
/home/niutr/pytorch/torch/lib/THD/base/data_channels/DataChannelMPI.cpp:194:34: error: conversion from ‘std::size_t {aka long unsigned int}’ to ‘at::Scalar’ is ambiguous
     output[i].copy_(recv_buffer[i]);
                                  ^
```

+1:

```
/home/geordi/projects/pytorch-master/torch/lib/THD/base/data_channels/DataChannelMPI.cpp: In member function ‘virtual void thd::DataChannelMPI::allGather(std::vector<at::Tensor>&, at::Tensor&, THDGroup)’:
/home/geordi/projects/pytorch-master/torch/lib/THD/base/data_channels/DataChannelMPI.cpp:194:34: error: conversion from ‘std::size_t {aka long unsigned int}’ to ‘at::Scalar’ is ambiguous
     output[i].copy_(recv_buffer[i]);
                                  ^
```

--- a/aten/src/ATen/TensorOperators.h
+++ b/aten/src/ATen/TensorOperators.h
@@ -57,6 +57,9 @@ inline Tensor Tensor::operator[](Tensor index) const {
   // The Scalar(Tensor) constructor is explicit, so we need to call it.
   return this->operator[](Scalar(index));
 }
+inline Tensor Tensor::operator[](size_t index) const {
+  return this->operator[](static_cast<int64_t>(index));
+}
 
 #define AT_FORALL_BINARY_OPS(_) \
 _(+,x.add(y), y.add(x)) \

--- a/aten/src/ATen/TensorOperators.h
+++ b/aten/src/ATen/TensorOperators.h
@@ -57,8 +57,8 @@ inline Tensor Tensor::operator[](Tensor index) const {
   // The Scalar(Tensor) constructor is explicit, so we need to call it.
   return this->operator[](Scalar(index));
 }
-inline Tensor Tensor::operator[](size_t index) const {
-  return this->operator[](static_cast<int64_t>(index));
+inline Tensor Tensor::operator[](int64_t index) const {
+  return select(0, index);
 }
 
 #define AT_FORALL_BINARY_OPS(_) \

--- a/aten/src/ATen/templates/Tensor.h
+++ b/aten/src/ATen/templates/Tensor.h
@@ -118,6 +118,7 @@ struct Tensor : public detail::TensorBase {
   Tensor& operator/=(Scalar other);
   Tensor operator[](Scalar index) const;
   Tensor operator[](Tensor index) const;
+  Tensor operator[](size_t index) const;
 
   // STOP.  Thinking of adding a method here, which only makes use
   // of other ATen methods?  Define it in native_functions.yaml.

--- a/aten/src/ATen/templates/Tensor.h
+++ b/aten/src/ATen/templates/Tensor.h
@@ -118,7 +118,7 @@ struct Tensor : public detail::TensorBase {
   Tensor& operator/=(Scalar other);
   Tensor operator[](Scalar index) const;
   Tensor operator[](Tensor index) const;
-  Tensor operator[](size_t index) const;
+  Tensor operator[](int64_t index) const;
 
   // STOP.  Thinking of adding a method here, which only makes use
   // of other ATen methods?  Define it in native_functions.yaml.

--- a/aten/src/ATen/test/basic.cpp
+++ b/aten/src/ATen/test/basic.cpp
@@ -272,6 +272,9 @@ static void test(Type & type) {
     for (int64_t i = 0; i < tensor.numel(); ++i) {
       ASSERT(tensor[i].equal(one * i));
     }
+    for (size_t i = 0; i < tensor.numel(); ++i) {
+      ASSERT(tensor[i].equal(one * static_cast<int64_t>(i)));
+    }
     for (int i = 0; i < tensor.numel(); ++i) {
       ASSERT(tensor[i].equal(one * i));
     }

